### DISCUSSION TITLE
KeyError when trying to iterate over MultiDiGraph.subgraph_monomorphisms_iter() with edge_match

### DISCUSSION THREAD

**Discussion by bramvile** [networkx/networkx#6851] [Q&A]
I have the following code:
```python
import networkx as nx
from networkx.algorithms.isomorphism import GraphMatcher, MultiDiGraphMatcher

G1 = nx.MultiDiGraph()
G1.add_nodes_from(
    [(0, {'label': 'a'}), (1, {'label': 'b'}), (2, {'label': 'b'})])
G1.add_edges_from([(0, 1)], relationship="enemy")
G1.add_edges_from([(0, 2), (2, 0)], relationship="friend")

SG1 = nx.MultiDiGraph()
SG1.add_nodes_from(
    [(3, {'label': 'a'}), (4, {'label': 'b'})])
SG1.add_edges_from([(3, 4), (4, 3)], relationship="friend")
monos = MultiDiGraphMatcher(G1, SG1, node_match=lambda a,
                            b: a['label'] == b['label'],
                            edge_match=lambda e1, e2: (print(e1),
                                                       e1['relationship'] == e2['relationship'])).subgraph_monomorphisms_iter()
for m in monos:
    print(m)
```

It results in the following error:
```
Traceback (most recent call last):
  File "/Users/danielbram/Programming/procgen-graph/server/main.py", line 200, in <module>
    for m in monos:
  File "/Users/danielbram/miniconda3/lib/python3.11/site-packages/networkx/algorithms/isomorphism/isomorphvf2.py", line 391, in subgraph_monomorphisms_iter
    yield from self.match()
  File "/Users/danielbram/miniconda3/lib/python3.11/site-packages/networkx/algorithms/isomorphism/isomorphvf2.py", line 316, in match
    yield from self.match()
  File "/Users/danielbram/miniconda3/lib/python3.11/site-packages/networkx/algorithms/isomorphism/isomorphvf2.py", line 313, in match
    if self.semantic_feasibility(G1_node, G2_node):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danielbram/miniconda3/lib/python3.11/site-packages/networkx/algorithms/isomorphism/vf2userfunc.py", line 168, in semantic_feasibility
    feasible = _semantic_feasibility(self, G1_node, G2_node)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/danielbram/miniconda3/lib/python3.11/site-packages/networkx/algorithms/isomorphism/vf2userfunc.py", line 64, in _semantic_feasibility
    if G2_nbr in G2nbrs and not edge_match(
                                ^^^^^^^^^^^
  File "/Users/danielbram/Programming/procgen-graph/server/main.py", line 194, in <lambda>
    e1['relationship'] == e2['relationship'])).subgraph_monomorphisms_iter()
    ~~^^^^^^^^^^^^^^^^
  File "/Users/danielbram/miniconda3/lib/python3.11/site-packages/networkx/classes/coreviews.py", line 53, in __getitem__
    return self._atlas[key]
           ~~~~~~~~~~~^^^^^
KeyError: 'relationship'
```

When I call `subgraph_monomorphisms_iter` without `edge_match`, I can iterate over `monos` with no issue. I logged the edges of both `G` and `SG1` with `.edges(data=True)` to confirm that all edges have an attribute called `relationship`, which they do.

Any idea what I might be doing wrong? I'm pretty new to Python so maybe I'm missing something that's not even related to `networkx` 🤷‍♂️ 

Thanks!

**Comment by dschult** [comment]
I think that error message is saying that `e1` does not have an attribute called `relationship`.  So, you'll either need to make every edge have a 'relationship' attribute, or change your `edge_match` function to use `e1.get('relationship', default)` instead of `e1['relationship']`.  You can leave off the `default` input to the `get` method and the default becomes `None`.  Hope this helps.

**Reply by bramvile** [reply]
@dschult Thank you for the quick and helpful response! It did solve the KeyError and the script runs, though I'm still confused as to why the previous method didn't work. I printed out the two graphs and this is what I see:
```python
print(list(SG1.edges(data=True)))
print(list(G1.edges(data=True)))

# [(3, 4, {'relationship': 'friend'})]
# [(0, 1, {'relationship': 'enemy'}), (0, 2, {'relationship': 'friend'}), (2, 0, {'relationship': 'friend'})
```

Seems like all edges have a `relationship` attribute, so I'm not sure which edge causes the KeyError.

**Reply by dschult** [reply]
Right -- I missed that and it's a good distinction. You **do** have a `relationshp` attribute for each edge. My suggestion of using `.get` will only fix edges that don't have the `relationship` attribute. And that is not your case.

The clue to solving your issue (I hope -- I may have missed something else) is to use your `print(e1)` output which appears before the error TraceBack. In your example it shows as:
```{0: {'relationship': 'friend'}}```
That's a clue that the dict is not just an attribute dict for a single edge. For the case of multigraphs, the edge_match input is the keydict representing all the edges between those two nodes.  So your `edge_match` function has to process the multiedges to determine whether there is a match between the node-pairs.  In other words, you need to figure out how you want to handle a case where one edge is a `friend` and another edge between the same two nodes is an `enemy`.  Perhaps something like:
```python
edge_match=lambda e1, e2: len({dd['relationship'] for ekey, dd in e1.items()} & {dd['relationship'] for ekey, dd in e2.items()}) > 0
```
or maybe better written as
```python
def any_edges_match(e1, e2):
    all_rel1 = {dd['relationship'] for ekeye, dd in e1.items()}
    all_rel2 = {dd['relationship'] for ekeye, dd in e2.items()}
    if all_rel1 & all_rel2:
        return True
    return False
```
